### PR TITLE
IGNITE-15322 Stop authorization of internal tasks by their class name.

### PR DESCRIPTION
**Pull request by petrov-mg**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/IgniteComputeImpl.java [hunk 1]
```diff
@@ -68,10 +68,11 @@ public class IgniteComputeImpl extends AsyncSupportAdapter<IgniteCompute>
     /** Custom executor name. */
     private String execName;
 
+    /** */
+    private boolean isSysTaskGuard;
```
Comment (anton-vinogradov) [inline modules/core/src/main/java/org/apache/ignite/internal/IgniteComputeImpl.java:72]: Could this be replaced with something explains it's a public/private api?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/task/TaskExecutionOptions.java [hunk 2]
```diff
@@ -176,13 +176,13 @@ public TaskExecutionOptions asSystemTask() {
...
     }
 
     /** */
-    public TaskExecutionOptions withAuthenticationDisabled() {
-        isAuthDisabled = true;
+    public TaskExecutionOptions withSystemTaskGuard() {
+        isSystemTaskGuard = true;
 
         return this;
     }
```
Comment (anton-vinogradov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/task/TaskExecutionOptions.java:188]: Looks like, this spetifies option for current thread only, is it safe?

Comment (petrov-mg) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/task/TaskExecutionOptions.java]: The option itself is not thread local. 
Take a look at IgniteComputeImpl, please. We do use TaskExecutionOptions as thread local variable here, but isPublicFacade flag is stored separately and applied to each compute execution method. So it is safe.

File: modules/core/src/main/java/org/apache/ignite/internal/IgniteComputeImpl.java (hunk 1 above)
Comment (petrov-mg) [reply modules/core/src/main/java/org/apache/ignite/internal/IgniteComputeImpl.java]: Fixed. Changed to isPublicFacade.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/task/GridTaskWorker.java [hunk 3]
```diff
@@ -1736,4 +1763,51 @@ else if (!getLocJobStatistics && locNodeId.equals(jobResult.getNode().id()))
...
+
+    /** */
+    public void authorizeTaskCancel() {
+        if (!ctx.security().enabled())
+            return;
+
+        if (!isSystemType(ctx, task.getClass())) {
+            ctx.security().authorize(task.getClass().getName(), TASK_CANCEL);
+
+            return;
```
[the remark this answers is not in the document]
Comment (petrov-mg) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/task/GridTaskWorker.java]: Fixed.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/task/GridTaskProcessor.java [hunk 4]
```diff
@@ -1589,4 +1579,36 @@ public Map<ComputeJobStatusEnum, Long> jobStatuses(IgniteUuid sesId) {
         else
             return taskWorker.jobStatuses();
     }
+
+    /** */
+    private void authorizeTaskExecution(String taskName, Class<?> taskCls, ComputeTask<?, ?> task) {
+        taskCls = resolveTaskClass(taskName, taskCls, task);
+
+        if (taskCls != null && SecurityUtils.isSystemType(ctx, taskCls))
+            return;
```
[the remark this answers is not in the document]
Comment (petrov-mg) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/task/GridTaskProcessor.java]: Fixed.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/task/GridTaskWorker.java [hunk 5]
```diff
@@ -1736,4 +1763,51 @@ else if (!getLocJobStatistics && locNodeId.equals(jobResult.getNode().id()))
...
     }
+
+    /** */
+    private void authorizeSystemJobStart(ComputeJob job) {
+        Object executable = unwrap(job);
+
+        if (!isSystemType(ctx, executable.getClass())) {
+            ctx.security().authorize(executable.getClass().getName(), TASK_EXECUTE);
+
+            return;
```
[the remark this answers is not in the document]
Comment (petrov-mg) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/task/GridTaskWorker.java]: Fixed.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/security/SecurityPermissionAware.java [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+/** */
+public interface SecurityPermissionAware {
+    /** */
+    public SecurityPermissionSet NO_PERMISSIONS = SecurityPermissionSetBuilder.create().build();
+
+    /** */
+    public SecurityPermissionSet requiredPermissions();
+
+   /** */
+    public static SecurityPermissionSet cachePermissions(String cacheName, SecurityPermission... perms) {
```
[the remark this answers is not in the document]
Comment (petrov-mg) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/security/SecurityPermissionAware.java]: Fixed.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/security/SecurityPermissionAware.java [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+package org.apache.ignite.internal.processors.security;
+
+import org.apache.ignite.plugin.security.SecurityPermission;
+import org.apache.ignite.plugin.security.SecurityPermissionSet;
+import org.apache.ignite.plugin.security.SecurityPermissionSetBuilder;
+
+/** */
+public interface SecurityPermissionAware {
+    /** */
+    public SecurityPermissionSet NO_PERMISSIONS = SecurityPermissionSetBuilder.create().build();
```
[the remark this answers is not in the document]
Comment (petrov-mg) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/security/SecurityPermissionAware.java]: Fixed.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/security/SecurityPermissionAware.java [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+ */
+
+package org.apache.ignite.internal.processors.security;
+
+import org.apache.ignite.plugin.security.SecurityPermission;
+import org.apache.ignite.plugin.security.SecurityPermissionSet;
+import org.apache.ignite.plugin.security.SecurityPermissionSetBuilder;
+
+/** */
+public interface SecurityPermissionAware {
```
[the remark this answers is not in the document]
Comment (petrov-mg) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/security/SecurityPermissionAware.java]: Fixed. The name was changed to SecurityAwareJob.

File: modules/core/src/main/java/org/apache/ignite/internal/cluster/ClusterGroupAdapter.java [hunk 9]
```diff
@@ -203,7 +203,7 @@ public void setKernalContext(GridKernalContext ctx) {
     /**
      * @return {@link IgniteCompute} for this cluster group.
      */
-    public final IgniteCompute compute() {
+    public final IgniteComputeImpl compute() {
```
[the remark this answers is not in the document]
Comment (petrov-mg) [reply modules/core/src/main/java/org/apache/ignite/internal/cluster/ClusterGroupAdapter.java]: Fixed.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/security/SecurityPermissionAware.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (petrov-mg) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/security/SecurityPermissionAware.java]: Fixed.

File: modules/core/src/main/java/org/apache/ignite/internal/cluster/ClusterGroupAdapter.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (petrov-mg) [reply modules/core/src/main/java/org/apache/ignite/internal/cluster/ClusterGroupAdapter.java]: Fixed.

File: modules/core/src/main/java/org/apache/ignite/internal/cluster/ClusterGroupAdapter.java [hunk 10]
```diff
@@ -201,16 +204,29 @@ public void setKernalContext(GridKernalContext ctx) {
     }
 
     /**
-     * @return {@link IgniteCompute} for this cluster group.
+     * @return Internal {@link IgniteCompute} facade for this cluster group.
+     */
+    public final IgniteCompute internalComputeFacade() {
```
[the remark this answers is not in the document]
Comment (petrov-mg) [reply modules/core/src/main/java/org/apache/ignite/internal/cluster/ClusterGroupAdapter.java]: Fixed.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/task/TaskExecutionOptions.java [hunk 11]
```diff
@@ -176,13 +176,13 @@ public TaskExecutionOptions asSystemTask() {
...
-    public boolean isAuthenticationDisabled() {
-        return isAuthDisabled;
+    public boolean isPublicRequest() {
+        return isPublicReq;
     }
 
     /** */
-    public TaskExecutionOptions withAuthenticationDisabled() {
-        isAuthDisabled = true;
+    public TaskExecutionOptions asPublicRequest() {
```
[the remark this answers is not in the document]
Comment (petrov-mg) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/task/TaskExecutionOptions.java:207]: Fixed.

File: modules/core/src/main/java/org/apache/ignite/internal/jdbc2/JdbcResultSet.java [hunk 12]
```diff
@@ -268,9 +268,10 @@ void closeInternal() throws SQLException {
                 JdbcConnection conn = (JdbcConnection)stmt.getConnection();
 
                 if (conn.isCloseCursorTaskSupported()) {
-                    Ignite ignite = conn.ignite();
+                    IgniteEx ignite = conn.ignite();
 
-                    ignite.compute(ignite.cluster().forNodeId(conn.nodeId())).call(new JdbcCloseCursorTask(uuid));
+                    ignite.internalCompute(ignite.cluster().forNodeId(conn.nodeId()))
+                        .call(new JdbcCloseCursorTask(uuid));
```
[the remark this answers is not in the document]
Comment (petrov-mg) [reply modules/core/src/main/java/org/apache/ignite/internal/jdbc2/JdbcResultSet.java]: Fixed.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/security/SecurityUtils.java [hunk 13]
```diff
@@ -385,4 +393,27 @@ public static SecurityContext authenticateLocalNode(
 
         return secCtx;
     }
+
+    /** */
+    public static void authorizeAll(IgniteSecurity security, SecurityPermissionSet permissions) {
+        if (!security.enabled())
+            return;
+
+        if (!F.isEmpty(permissions.systemPermissions())) {
```
[the remark this answers is not in the document]
Comment (petrov-mg) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/security/SecurityUtils.java]: Fixed.

File: modules/clients/src/test/java/org/apache/ignite/common/ComputeTaskPermissionsTest.java [hunk 14]
```diff
@@ -0,0 +1,624 @@
...
+    }
+
+    /** */
+    private static class AbstractTask extends ComputeTaskAdapter<Object, Object> {
+        /** {@inheritDoc} */
+        @Override public @NotNull Map<? extends ComputeJob, ClusterNode> map(
+            List<ClusterNode> subgrid,
+            @Nullable Object arg
+        ) throws IgniteException {
+            return subgrid.stream().filter(g -> !g.isClient()).collect(Collectors.toMap(g -> job(), g -> g, (l, r) -> l));
```
[the remark this answers is not in the document]
Comment (petrov-mg) [reply modules/clients/src/test/java/org/apache/ignite/common/ComputeTaskPermissionsTest.java]: Fixed.

**Review by anton-vinogradov — APPROVED** [review]

**Pull request merged**